esp-hal/src/gpio/interrupt.rs

```diff
@@ -0,0 +1,154 @@
+use portable_atomic::{AtomicPtr, Ordering};
```

Comment:
I feel like some docs here somewhere outlining the contract between the interrupt handler and the pins would be helpful for future us.

For example, there are assumptions in `Drop for PinFuture` about the interrupt being already disabled. I think we should be really clear about the responsibilities of the interrupt handler, and the assumptions for pin/pinfutures.